File: hash_table/hash_table.py
class Node:
    def __init__(self, key, value):
        self.key = key
        self.value = value
        self.next = None


class HashTable:
    def __init__(self, initial_capacity=8, load_factor=0.7):
        self.capacity = initial_capacity
        self.size = 0
        self.table = [None]*initial_capacity
        self.load_factor = load_factor

    def __str__(self):
        elements = []
        for i in range(self.capacity):
            current = self.table[i]
            while current:
                elements.append((current.key, current.value))
                current = current.next
        return str(elements)

    def __len__(self):
        return self.size

    def __contains__(self, key):
        try:
            self.search(key)
            return True
        except KeyError:
            return False

    def _hash(self, key):
        return hash(key) % self.capacity

    def _resize(self):
        old_table = self.table
        self.capacity *= 2
        self.size = 0
        self.table = [None]*self.capacity

        for node in old_table:
            current = node
            while current:
                self.insert(current.key, current.value)
                current = current.next

    def insert(self, key, value):
        if self.size / self.capacity > self.load_factor:
            self._resize()

        index = self._hash(key)

        if self.table[index] is None:
            self.table[index] = Node(key, value)
            self.size += 1
        else:
            current = self.table[index]
            while current:
                if current.key == key:
                    current.value = value
                    return
                current = current.next

            # Else create a node and insert it at the head
            new_node = Node(key, value)
            new_node.next = self.table[index]
            self.table[index] = new_node
            self.size += 1

    def search(self, key):
        index = self._hash(key)
        current = self.table[index]

        while current:
            if current.key == key:
                return current.value
            current = current.next
        raise KeyError(key)

    def remove(self, key):
        index = self._hash(key)
        previous = None
        current = self.table[index]

        while current:
            if current.key == key:
                if previous:
                    previous.next = current.next
                else:
                    self.table[index] = current.next
                    self.size -= 1
                    return
            previous = current
            current = current.next

        raise KeyError(key)


def test_hash_table():
    ht = HashTable(4)
    print(ht)

    ht.insert("apple", 3)
    ht.insert("banana", 2)
    print(ht)
    assert ht.capacity == 4

    ht.insert("cherry", 5)
    ht.insert("blueberry", 5)
    print(ht)
    size = 4
    assert len(ht) == size
    assert ht.capacity == 8

    assert "apple" in ht
    assert "durian" not in ht
    assert ht.search("banana") == 2

    ht.insert("banana", 4)
    print(ht)
    assert ht.search("banana") == 4
    assert len(ht) == size

    ht.remove("apple")
    size -= 1
    print(ht)
    assert len(ht) == size


if __name__ == "__main__":
    test_hash_table()
    print("All tests passed !")

File: hash_table/hash_table.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include "hash_table.h"

Node *create_node(const char *key, int value) {
    Node *new_node = (Node *)malloc(sizeof(Node));
    new_node->key = strdup(key);
    new_node->value = value;
    return new_node;
}

HashTable *create_table(int initial_capacity, float load_factor) {
    HashTable *ht = (HashTable *)malloc(sizeof(HashTable));
    ht->capacity = initial_capacity;
    ht->size = 0;
    ht->load_factor = load_factor;
    ht->table = (Node **)calloc(initial_capacity, sizeof(Node *));
    return ht;
}

void free_table(HashTable *ht) {
    for (int i = 0; i < ht->capacity; i++) {
        Node *current = ht->table[i];
        while (current) {
            Node *temp = current;
            current = current->next;
            free(temp->key);
            free(temp);
        }
    }
    free(ht->table);
    free(ht);
}

void print_table(HashTable *ht){
    printf("[");
    int first = 1;
    for (int i=0; i < ht->capacity; i++) {
        Node *current = ht->table[i];
        while (current) {
            if (!first) {
                printf(", ");
            } else {
                first = 0;
            }
            printf("(%s, %d)", current->key, current->value);
            current = current->next;
        }
    }
    printf("]\n");
}

// djb2 hash function
unsigned long hash(const char *key, int capacity)  {
    unsigned long hash = 5381;
    int c;
    while ((c = *key++)) {
        hash = ((hash << 5) + hash) + c;
    }
    return hash % capacity;
}

void insert(HashTable *ht, const char *key, int value) {
    if ((float)ht->size / ht->capacity > ht->load_factor) {
        resize(ht);
    }

    int index = hash(key, ht->capacity);
    Node *current = ht->table[index];

    while (current) {
        if (strcmp(current->key, key) == 0) {
            current->value = value;
            return;
        }
        current = current->next;
    }

    // create node at the head of the chained array
    Node *new_node = create_node(key, value);
    new_node->next = ht->table[index];
    ht->table[index] = new_node;
    ht->size++;
}

void resize(HashTable *ht) {
    int old_capacity = ht->capacity;
    Node **old_table = ht->table;

    ht->capacity *= 2;
    ht->size = 0;
    ht->table = (Node **)calloc(ht->capacity, sizeof(Node *));

    for (int i = 0; i < old_capacity; i++) {
        Node *current = old_table[i];
        while (current) {
            insert(ht, current->key, current->value);
            Node *temp = current;
            current = current->next;
            free(temp->key);
            free(temp);
        }
    }
    free(old_table);
}

int search(HashTable *ht, const char *key) {
    int index = hash(key, ht->capacity);
    Node *current = ht->table[index];

    while (current) {
        if (strcmp(current->key, key) == 0) {
            return current->value;
        }
        current = current->next;
    }

    fprintf(stderr, "KeyError: %s not found\n", key);
    exit(EXIT_FAILURE);
}

void remove_key(HashTable *ht, const char *key) {
    int index = hash(key, ht->capacity);
    Node *current = ht->table[index];
    Node *previous = NULL;

    while (current) {
        if (strcmp(current->key, key) == 0) {
            if (previous) {
                previous->next = current->next;
            } else {
                ht->table[index] = current->next;
            }
            free(current->key);
            free(current);
            ht->size--;
            return;
        }
        previous = current;
        current = current->next;
    }

    fprintf(stderr, "KeyError: %s not found\n", key);
    exit(EXIT_FAILURE);
}

void test_hash_table() {
    HashTable *ht = create_table(4, 0.7);
    print_table(ht);
    printf("Capacity: %d\n", ht->capacity);

    insert(ht, "apple", 3);
    insert(ht, "banana", 2);
    insert(ht, "cherry", 5);
    insert(ht, "blueberry", 5);
    
    print_table(ht);
    printf("Capacity after insertion: %d\n", ht->capacity);

    int size = 4;
    if (ht->size != size){
        fprintf(stderr, "Size mismatch, expected %d, got %d\n", size, ht->size);
    }

    if (search(ht, "apple") == 3) {
        printf("Found apple: %d\n", search(ht, "apple"));
    } else {
        fprintf(stderr, "apple not found\n");
    }

    if (search(ht, "banana") == 2) {
        printf("Found banana: %d\n", search(ht, "banana"));
    } else {
        fprintf(stderr, "banana not found\n");
    }

    insert(ht, "banana", 4);
    print_table(ht);
    if (search(ht, "banana") == 4) {
        printf("Updated banana: %d\n", search(ht, "banana"));
    } else {
        fprintf(stderr, "banana not found after update\n");
    }

    remove_key(ht, "apple");
    size--;
    print_table(ht);
    if (ht->size != size) {
        fprintf(stderr, "Size mismatch after removal: expected %d, got %d\n", size, ht->size);
    }
    free_table(ht);
}

int main() {
    test_hash_table();
    printf("All tests passed!\n");
    return 0;
}
